GHzDACs/ghz_fpga_server.py

```diff
@@ -1644,18 +1659,66 @@ def dac_run_sram(self, c, data, loop=False, blockDelay=0):
         parameters specifies the number of microseconds to delay
         for a multiblock sequence.
         """
-        if not len(data):
-            return
-        
-        if loop:
-            # make sure data is at least 20 words long by repeating it
-            data *= (20-1)/len(data) + 1
-        else:
-            # Set data at least 20 words long by repeating first value
-            data = list(data) + [data[0]] * (20-len(data))
-        
+        if len(data) < 20:
+            raise ValueError("Cannot play less than 20 ns of data.")
+
         dev = self.selectedDAC(c)
         yield dev.runSram(data, loop, blockDelay)
+
+    @setting(2081, 'DAC Write SRAM', data='*w')
+    def dac_write_sram(self, c, data):
+        """ Write the given data to SRAM.
+
+        Note that if your board is currently running, it will immediately start
+        playing the new SRAM, wherever it is, so use with caution.
+        """
+        dev = self.selectedDAC(c)
+        yield dev._sendSRAM(np.array(data, dtype='<u4').tostring())
+
+    @setting(1082, "Jump Table Add Entry", name='s', arg=['ww{IDLE}', 'www{JUMP}', 'w{NOP,END}', 'wwww{CYCLE}'])
+    def jump_table_add_entry(self, c, name, arg=None):
+        """ Add a new jump table entry.
+
+        Entry types are: IDLE, JUMP, NOP, END, CYCLE, CHECK, RAMP
+        (as of now, CHECK and RAMP are unimplemented)
+
+        The arguments vary depending on the command, see the JT docs for
+        details. Note that from addresses and to addresses are to be specified
+        in nanoseconds (though still in words). They will be divided by 4 to
```

Comment:
Specifying the from addresses in integer numbers of nanoseconds seems dangerous. What happens if I pick `3`? No error is raised here, and it's not even clear what *should* happen.